panel/layout/log.py

```diff
@@ -0,0 +1,159 @@
+from __future__ import annotations
+
+from typing import (
+    TYPE_CHECKING, ClassVar, List, Mapping, Optional, Type,
+)
+
+import param
+
+from ..models import Log as PnLog
+from ..models.log import ScrollButtonClick
+from .base import Column
+
+if TYPE_CHECKING:
+    from bokeh.document import Document
+    from bokeh.model import Model
+    from pyviz_comms import Comm
+
+    from ..viewable import Viewable
+
+
+class Log(Column):
+
+    load_buffer = param.Integer(default=50, bounds=(0, None), doc="""
+        The number of objects loaded on each side of the visible objects.
+        When scrolled halfway into the buffer, the log will automatically
+        load additional objects while unloading objects on the opposite side.""")
+
+    scroll = param.Boolean(default=True, doc="""
+        Whether to add scrollbars if the content overflows the size
+        of the container.""")
+
+    visible_indices = param.List(constant=True, doc="""
```

Comment:
readonly: cannot be set, ever
constant: can only be set on instantiation

```suggestion
    visible_indices = param.List(readonly=True, doc="""
```

Reply:
Maybe we should change this to `visible_range` after all (but leave the `visible_objects` bokeh property the same).

Reply:
I need to set it though to handle the scroll button I think:
https://github.com/holoviz/panel/pull/6031/files#diff-9420fdc4f24f25791a1047e728c7ab27e328e6bc0fec1a56427e9d57d03d2139R151-R155

Reply:
It's not synced with JS, so you'd simply update _synced_indices.

Reply:
I might be misunderstanding and missing something.

If I do have a `_synced_indices.setter`, wouldn't all this logic be gone?
```python
    @property
    def _synced_indices(self):
        n = len(self.objects)
        if self.visible_indices:
            return list(range(
                max(min(self.visible_indices) - self.load_buffer, 0),
                min(max(self.visible_indices) + self.load_buffer, n)
            ))
        elif self.view_latest:
            return list(range(max(n - self.load_buffer * 2, 0), n))
        else:
            return list(range(min(self.load_buffer, n)))
```

If I do `self._custom_synced_indices`

```python

    @property
    def _synced_indices(self):
        if self._custom_synced_indices is not None:
            return self._custom_synced_indices

        n = len(self.objects)
        if self.visible_indices:
            return list(range(
                max(min(self.visible_indices) - self.load_buffer, 0),
                min(max(self.visible_indices) + self.load_buffer, n)
            ))
        elif self.view_latest:
            return list(range(max(n - self.load_buffer * 2, 0), n))
        else:
            return list(range(min(self.load_buffer, n)))
```

```python

        self._custom_synced_indices = list(
            range(max(n - len(self.visible_indices) + 1, 0), n)
        )
        self.param.trigger("objects")
        self._custom_synced_indices = None
```

There seems to be some race condition and it unsuccessfully scrolls.